### PR TITLE
Mazda CX-9 2021 no steer lockout and more torque

### PR DESCRIPTION
**Pull request by Jafaral**
New Mazda updates:

   Mazda CX-9 2021 does not have steering lockout
   Mazda CX-9 2019 and newer have double the torque we have seen in earlier years/models
   Mazda CX-5 2021 still has lockout, but the timeout was increased to 30 seconds up form 5

  Mazda CX-9 has been added to readme as a supported car. It doesn't need a new fingerprint.
  Old test routes from the initial port still apply.

Signed-off-by: Jafar Al-Gharaibeh <to.jafar@gmail.com>

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
We'll need to be able to differentiate the CX-9 without the lockout if we want to move it out of dashcam only, so let's make it a separate car.

File: selfdrive/car/mazda/values.py [hunk 1]
```diff
@@ -8,13 +8,13 @@
 # Steer torque limits
 
 class CarControllerParams:
-  STEER_MAX = 600                 # max_steer 2048
-  STEER_STEP = 1                  # how often we update the steer cmd
+  STEER_MAX = 1500                # theoretical max_steer 2047
```
Comment (adeebshihadeh) [inline selfdrive/car/mazda/values.py:11]: why was this raised? is this valid for all cars?

**Comment by Jafaral** [comment]
> We'll need to be able to differentiate the CX-9 without the lockout if we want to move it out of dashcam only, so let's make it a separate car.

Unfortunately there is no easy way to do that currently.  All CX-5 and CX-9 models 2018+ share the same fingerprint. The few fingerprints for CX-9 we have are old/Australian models.

**Comment by pd0wm** [comment]
Do any of the current FW queries return anything for Mazda? This might be the time to figure out ECU FW fingerprinting.

File: selfdrive/car/mazda/values.py (hunk 1 above)
Comment (Jafaral) [reply selfdrive/car/mazda/values.py:11]: Hi @adeebshihadeh, instead of pumping this to 1500, I will collect some data on the behavior at different values and select the max value that works best. I will update this PR when I get a chance.

Comment (Jafaral) [reply selfdrive/car/mazda/values.py:11]: The EPS in older cars max at 308, the new ones at 620. In my testing with older cars I selected 600 based on testing. Even though the EPS will trim values above 308, it responds better to higher values by ramping up the torque faster. Reviewing drive data from newer cars show that the don't have the ramp up delay we had in older cars. so a value in the range 620-800 would work best across the board.

**Comment by Jafaral** [comment]
@pd0wm, I haven't tested recently but I did get some results last time I checked, and I already started working on FP 2.0, I had some missing pieces and I didn't get a chance to finish it.

**Comment by pd0wm** [comment]
Cool! Can you give some more details?

 - What exact UDS query is your car responding to?
 - What ECUs respond?
 - What do the responses look like?
 - How do the responses compare between different models/year.

**Comment by Jafaral** [comment]
@pd0wm 

```Getting VIN took 0.112 s

  0%|                                                                                    | 0/3 [00:00<?, ?it/s]iso-tp query bad response: 0x7f2231
iso-tp query bad response: 0x7f2231
iso-tp query bad response: 0x7f2231
iso-tp query bad response: 0x7f2231
iso-tp query bad response: 0x7f2231
iso-tp query bad response: 0x7f2231
iso-tp query bad response: 0x7f2231
iso-tp query bad response: 0x7f3e13
iso-tp query bad response: 0x7f3e13
iso-tp query bad response: 0x7f2231
100%|████████████████████████████████████████████████████████████████████████████| 3/3 [00:02<00:00,  1.35it/s]

Found FW versions
{
  (Ecu.eps, 0x730, None): [b'KJ01-3210X-G-00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.engine, 0x7e0, None): [b'PYNF-188K2-F\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.fwdRadar, 0x764, None): [b'K123-67XK2-F\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.esp, 0x760, None): [b'K123-437K2-E\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.fwdCamera, 0x706, None): [b'B61L-67XK2-T\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
}

Possible matches: {'Mazda CX-5 2017'}
Getting fw took 2.265 s
```

**Comment by Jafaral** [comment]
New PR at #20879

**Comment by adeebshihadeh** [comment]
Converting to a draft until #20879 is merged

File: selfdrive/car/mazda/values.py [hunk 2]
```diff
@@ -8,18 +8,19 @@
...
+  STEER_ERROR_MAX = 350           # max delta between torque cmd and torque motor
 
 class CAR:
-  CX5 = "Mazda CX-5 2017"
-  CX9 = "Mazda CX-9 2017"
-  Mazda3 = "Mazda3 2017"
+  CX5 = "Mazda CX-5"
+  CX9 = "Mazda CX-9"
+  Mazda3 = "Mazda 3"
+  CX9_NSL = "Mazda CX-9 2021"   # No Steer Lockout
```
Comment (adeebshihadeh) [inline selfdrive/car/mazda/values.py:23]: call this CX9_2021. you could also call it CX9 and the older one CX9_YEAR

File: selfdrive/car/mazda/interface.py [hunk 3]
```diff
@@ -21,7 +21,8 @@ def get_params(candidate, fingerprint=gen_empty_fingerprint(), car_fw=None):
     ret.carName = "mazda"
     ret.safetyModel = car.CarParams.SafetyModel.mazda
 
-    ret.dashcamOnly = True
+    if candidate in STEER_LOCK_CAR:
+      ret.dashcamOnly = True
```
Comment (adeebshihadeh) [inline selfdrive/car/mazda/interface.py:25]:
```suggestion
    ret.dashcamOnly = candidate in STEER_LOCK_CAR
```

**Comment by Jafaral** [comment]
Thanks @adeebshihadeh. I made the changes.

I will work on getting routes for these two new cars tp pass the test.

File: selfdrive/car/mazda/values.py [hunk 4]
```diff
@@ -8,18 +8,20 @@
...
   STEER_DRIVER_ALLOWANCE = 15     # allowed driver torque before start limiting
   STEER_DRIVER_MULTIPLIER = 1     # weight driver torque
   STEER_DRIVER_FACTOR = 1         # from dbc
+  STEER_ERROR_MAX = 350           # max delta between torque cmd and torque motor
 
 class CAR:
-  CX5 = "Mazda CX-5 2017"
-  CX9 = "Mazda CX-9 2017"
-  Mazda3 = "Mazda3 2017"
+  CX5 = "Mazda CX-5"
```
Comment (pd0wm) [inline selfdrive/car/mazda/values.py:20]: Can we make the car names match the other brands? Keys and values should be all caps, so `MAZDA3 = MAZDA 3`. Would be nice if you can also find a year or generation name for the older cars.

Comment (Jafaral) [reply selfdrive/car/mazda/values.py]: I did  change the name. I think I'm going to hold on giving a generation name for now until we see a new generation upstreams. I already refer to these cars as "Gen1" Mazda ADAS system which include most Mazda cars so far. the new Mazda3 and CX-30 have a new system, we will call that Gen2 when it is supported.

**Comment by Jafaral** [comment]
This PR is in good shape, The last piece is just getting a  route to add to the tests. I'm still waiting to get one.

File: selfdrive/car/mazda/values.py [hunk 5]
```diff
@@ -114,17 +128,66 @@ class Buttons:
       b'B61L-67XK2-T\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
     ],
     (Ecu.transmission, 0x7e1, None): [
+      b'PY2S-21PS1-C\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
       b'P52G-21PS1-F\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
       b'PYKE-21PS1-B\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
     ],
+  },
+
+  CAR.MAZDA6: {
```
Comment (adeebshihadeh) [inline selfdrive/car/mazda/values.py:137]: this should go in a separate PR

File: selfdrive/car/mazda/values.py [hunk 6]
```diff
@@ -52,49 +55,59 @@ class Buttons:
     ],
     (Ecu.esp, 0x760, None): [
       b'K123-437K2-E\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
+      b'KL2K-437K2-A\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
       b'KN0W-437K2-C\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
       b'KBJ5-437K2-B\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
     ],
     (Ecu.fwdCamera, 0x706, None): [
       b'B61L-67XK2-T\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
+      b'B61L-67XK2-R\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
```
Comment (adeebshihadeh) [inline selfdrive/car/mazda/values.py:64]: anything not related to the CX9 port should go in separate PRs

Comment (Jafaral) [reply selfdrive/car/mazda/values.py:64]: I can do that, but since there wasn't really any major work I thought it be fine to just stick it in here.

**Comment by Jafaral** [comment]
@adeebshihadeh , I did Mazda6 and other changes to #21167. Once we merge that one I will rebase this Mazda CX9 2021 PR.

**Comment by adeebshihadeh** [comment]
I think it's best to get this merged as dashcam only until we can get a CX-9 to verify the port.

**Review by Jafaral — COMMENTED** [review]
@adeebshihadeh , I rebased and moved CX-9 2021 to the list of cars with steer lockout until it gets tested

File: README.md [hunk 7]
```diff
@@ -175,6 +175,7 @@ Community Maintained Cars and Features
 | Kia       | Sorento 2018-19               | SCC + LKAS        | Stock            | 0mph               | 0mph         |
 | Kia       | Stinger 2018                  | SCC + LKAS        | Stock            | 0mph               | 0mph         |
 | Kia       | Ceed 2019                     | SCC + LKAS        | Stock            | 0mph               | 0mph         |
+| Mazda     | CX-9 2021                     | i-Activesense     | Stock            | 0mph               | 32mph        |
```
Comment (adeebshihadeh) [inline README.md:178]: remove until it's out of dashcam only
```suggestion
```

File: selfdrive/car/mazda/interface.py [hunk 8]
```diff
@@ -21,7 +21,7 @@ def get_params(candidate, fingerprint=gen_empty_fingerprint(), car_fw=None):
     ret.carName = "mazda"
     ret.safetyModel = car.CarParams.SafetyModel.mazda
 
-    ret.dashcamOnly = True
+    ret.dashcamOnly = candidate in STEER_LOCK_CAR
```
Comment (adeebshihadeh) [inline selfdrive/car/mazda/interface.py:24]: let's also set the community feature flag for all models since the CX9 will need to be behind it anyway

**Comment by adeebshihadeh** [comment]
Can you add a test route?

**Comment by adeebshihadeh** [comment]
Any update on this?

**Comment by Jafaral** [comment]
I asked the person to provide a drive and still waiting for one.
